### PR TITLE
Welt Online.js -- fix eslint warnings and re-license

### PR DESCRIPTION
**Pull request by retorquere**
Is GPL-licensed, not AGPL

### CODE REVIEW

File: Welt Online.js [hunk 1]
```diff
@@ -151,5 +151,5 @@ var testCases = [
 		"url": "http://www.welt.de/wirtschaft/",
 		"items": "multiple"
 	}
-]
-/** END TEST CASES **/
\ No newline at end of file
+];
```
[the remark this answers is not in the document]
Comment (retorquere) [reply Welt Online.js]: The linter from https://github.com/zotero/translators/pull/1895 wouldn't do this. I'd prefer to wait until that drops and then revisit these PRs.

[the remark this answers is not in the document]
Comment (zuphilip) [reply Welt Online.js]: Okay.

[the remark this answers is not in the document]
Comment (retorquere) [reply Welt Online.js]: Is it actually an error for there to be a `;` at the end of the testcases BTW? Just curious.

[the remark this answers is not in the document]
Comment (zuphilip) [reply Welt Online.js]: If you treat this as JSON then a `;` is wrong. However, I guess it is still running in most places.

My main reason is that we should not change it here, and then next in Scaffold change it back again etc.

[the remark this answers is not in the document]
Comment (retorquere) [reply Welt Online.js]: Not a problem.

**Comment by zuphilip** [comment]
@retorquere What is the status of these PRs?

**Comment by retorquere** [comment]
These PRs only fix linter errors, they make no functional changes. I think Dan prefers only linter fixes to be introduced when a translator changes functionally; if that's the case, all my "fix eslint warnings" PRs can be closed.

**Comment by adam3smith** [comment]
@dstillman -- I'd actually like to take the PRs for which @retorquere has already done the linting work (agree on not doing any new "just linting fixes" ones). It's relatively fast to fix linting, but still a non-trivial effort and will save time down the road. Would that be OK? If not, I'd indeed want to close these to have the list of open PRs be more manageable again.

**Comment by dstillman** [comment]
Fine with me. Thanks to both of you for going through these.

**Comment by retorquere** [comment]
In which case, let me make sure each of these adheres to the newest linting rules. New ones were added to @zotero/linting since I opened those PRs.

**Comment by zuphilip** [comment]
@retorquere And make sure that the test cases were left untouched (no semi-colon at the end) as mentioned above.

**Comment by retorquere** [comment]
How do you guys want me to mark the ones that I've inspected? I can add "updated to @zotero/eslint-config@1.0.2" in the subject for example.

**Comment by retorquere** [comment]
Is the return value of the callback in `Zotero.selectItems` used? I don't see it being used in the Zotero code, but I see translators doing `return true` for early termination but no `return false` at the end (which implies `return undefined`, but I don't know whether the `return true` should just be `return`, or whether I should add `return false` at the end`)

**Comment by dstillman** [comment]
Yeah, I actually don't know — I was discussing that with @zuphilip the other day.

https://github.com/zotero/translators/pull/1950#issuecomment-498303261

**Comment by zuphilip** [comment]
I went with `return;` which worked and seemed for me the most natural option to choose.

**Comment by zuphilip** [comment]
For marking the updated PR, I would just suggest that you make a comment like "Updated" in the PR, which we will see.

**Comment by retorquere** [comment]
> I went with `return;` which worked and seemed for me the most natural option to choose.

And as far as I can tell, the callback return value is not used. I'll follow the same.

> For marking the updated PR, I would just suggest that you make a comment like "Updated" in the PR, which we will see.

As a new comment then? Alright.

**Comment by retorquere** [comment]
And ZU.doGet? I'm seeing inconsistent returns there too.

**Comment by dstillman** [comment]
Definitely no return value there

**Comment by retorquere** [comment]
updated for @zotero/eslint-config@1.0.2 (license warning remains)

**Comment by zuphilip** [comment]
About the GPL license here: I am fine with AGPL for my part of the translator here, which might be more than the original code. But relicensing in general is a tricky business and I stopped an earlier attempt in this direction.

**Comment by retorquere** [comment]
I see these names through `git blame`:

* Avram Lyon
* Emiliano Heyns
* Philipp Zumstein
* Simon Kornblith
* adam3smith 

But that's not to say no other contributors have existed whose names git erased by someone else correcting a typo.

**Comment by zuphilip** [comment]
I just checked the old PR https://github.com/zotero/translators/issues/951 and it seems that all contributors you listed should be fine with AGPL. However, the [initial commit message](https://github.com/zotero/translators/commit/5fa0d5af7ac5a079bf22ad1aa226f3b08df732ab#diff-71c4e99441ec74352742f951ec07d5d1) tell us that this translator comes from Martin Meyerhoff, which is also the one mentioned in the current GPL license here. @dstillman Any suggestions?

**Comment by dstillman** [comment]
@mamhoff: Would you be OK with relicensing your GPL Zotero translators from 2011 as AGPL?

**Comment by mamhoff** [comment]
@dstillman thank you for asking, and for digging my account up! Yes, please do relicense as AGPL. Good move imo!

**Comment by zuphilip** [comment]
@mamhoff Thank you very much for the fast and positive reply! We will change the license to AGPL then.

**Comment by zuphilip** [comment]
I sugest we do here two commits in the end: one for fixing all the linting issues, and another one for changing the license. @dstillman Do you agree?

**Comment by retorquere** [comment]
(oh no please not rebase 😢 )

**Comment by dstillman** [comment]
I'd strongly suggest just spending some time with some interactive-rebase documentation — it really need not be scary, and it's a critical tool for building clean histories — but in a case like this you can also just `reset [commit before the first change]` and then `add -p` each chunk (using `s` to split if necessary) to create two separate commits, and then force push.

**Comment by retorquere** [comment]
I currently have about a 50% success rate between getting a rebase that worked and screwing it up to the point that it's easier to start over.

Would a force-push achieve the same? I could just re-do everything and apply the re-license last.

**Comment by retorquere** [comment]
What does `add -p` do?

**Comment by dstillman** [comment]
> I currently have about a 50% success rate between getting a rebase that worked and screwing it up to the point that it's easier to start over.

That's why I'm suggesting reading more documentation on it. It's a standard part of a git workflow.

If linting changes and real changes are separated in commits here, rebase would let you combine the relevant commits into two commits in about 10 seconds.

If you want to reassemble the commits from scratch, you can follow my alternative workflow above.

**Comment by dstillman** [comment]
I'll acknowledge that some rebase documentation does get overly into graph theory and confusing diagrams.

In a case like this, the upshot is that `git rebase -i upstream/master` would show you the list of commits unique to this branch, the same as what you see in the Commits tab (though without the pointless merge commits). You'd leave two lines as `p` (for pick`) and change the others to `s` (for squash) and reorder them so they're squashed into the right `p` lines. You'd then save and quit the editor, and it would then prompt you to edit the two final commit messages into clean messages. After that, `git log` should show two commits since the upstream/master commit, and you would force-push those to update this PR for merging.

This sounds complicated, but it's really quite simple in practice once you get the hang of it.

**Comment by retorquere** [comment]
There isn't really any "real changes" here apart from the relicensing. Please do check whether I got it right -- man I hate rebase.

**Comment by dstillman** [comment]
No, that’s not what fixup is for — that would squash all but the first commit. If you want to rebase on the current master you would just make sure your local master has been fetched (via `fetch` or via a `pull` with that branch checked out) and then run the rebase again without making any changes. But we can do that via the PR at merge time (when we don’t want to create a merge commit), so that’s not necessary if there aren’t any changes from master that you know you need.

The branch point is just the commit before the first commit for this branch in git log.

**Comment by mamhoff** [comment]
This thread makes me super-happy. When these files were first committed, @avram committed them for me because I did not know how git works at all. 

They were my first-ever open source contribution, and they actually got me recognition and possibilities. Shout out to you, Avram! 

As for learning to rebase and all the interesting git-fu I can very much recommend https://learngitbranching.js.org/ - enjoy! 

❤️ to all of you, keep up the good work, and thank you again!

**Comment by retorquere** [comment]
I know a little about graph theory, and I prefer classes in graph theory of docs on rebasing. The docs I found so far seem to be all to happy to go into all the esoteric possibilities rather than having a section on "you have a simple branch off of master, you want to squash the first N commits into one" scenario. Meanwhile, history rewrite runs counter to anything I have learned about version control, and git, being git, will happily allow you to screw the history beyond sensible recovery, which is absolutely not what a VC tool should be doing.

I understand the desire for a cleaner history, and so does github because I reckon that's why they offer squash & merge.

I've found a GUI that makes it slightly less terrifying, but would it for these kinds of things not be possible to have two local branches, one with the functional changes, and one deriving from it with the lint fixes? And then offer these bundled in some way? Or just as two separate PRs? It's easy enough for me to have one branch with functional changes and another branch derived from that with functional changes + linting applied. I've always used branches for things that I wan't to be able to judge separately.

Or I can learn to live with rebase of course. But rebase is sucking the joy out of what is otherwise a happy job for me.

**Comment by dstillman** [comment]
Thanks, @mamhoff! That's great to hear.

**Comment by dstillman** [comment]
@retorquere:

> git, being git, will happily allow you to screw the history beyond sensible recovery

Not really. You can always do a hard reset to a previous commit from the reflog or from GitHub (and you should never push (let alone force-push) changes without reviewing them locally first, so there shouldn't any doubt about what you're pushing). As long as you haven't force-pushed, you can even always throw out the local branch and pull it again from GitHub.

Beyond that, I don't really know what to tell you. Once you get the hang of it, an interactive rebase is really quite simple and fast. If you use `commit --fixup` and turn on autosquash, you don't even need to make any changes during the rebase. And `reset` + `add -p` is always an alternative, super-easy approach, albeit a slower one. (Frankly I find all Git GUIs to be far scarier and error-prone.)

But one way or another, we need these different changes to be in separate commits, and trying to start reconciling separate branches isn't going to make anything easier. If you don't want to work on these, that's really no problem.

**Comment by retorquere** [comment]
There's no point in me moaning about rebase really. Rebase is what is required to get changes into Zotero. I'll do my best to not bring it up again.

I've done the rebase for this PR to the best of my abilities. There are now two commits here, which I think split off the lint and the license changes. If there's no order preference between them, I'd prefer for future PRs to do the functional change as the first commit, and the lint as the last.

**Comment by zuphilip** [comment]
@retorquere Maybe this video helps: https://bibserv24.bib.uni-mannheim.de/cloud/index.php/s/a5irjCbSYPkXRo6 (there might be several shortcuts to what I did there...)

**Pull request merged**